dockeradapter/filterdockerstream.go

```diff
@@ -40,46 +47,67 @@ const (
-		// and empty slice?
-		bytesReturned int
-	)
-	_, err := reader.Read(prefix)
+	logging.Stderr("[FilterDockerstream] filtering reader for steam type: %d", streamType)
+	defer logging.Stderr("[FilterDockerstream] leaving")
+	result := []string{}
+	h, err := readFrameHeader(reader)
 	for err == nil {
-		payloadLength = int(binary.BigEndian.Uint32(prefix[stdWriterSizeIndex:]))
-		payload := make([]byte, payloadLength)
-		bytesReturned, err = reader.Read(payload)
+		payload, err := readFramePayload(reader, h)
 		if err != nil {
-			return nil, fmt.Errorf("Error reading payload")
-		}
-		if bytesReturned != payloadLength {
-			return nil,
-				fmt.Errorf("Bytes returned from stream [%d] does not match the length specified in the header [%d]",
-					bytesReturned,
-					payloadLength,
-				)
+			logging.Stderr("[FilterDockerstream] error reading payload: %s", err)
+			return result, err
 		}
-		lines := bytes.Split(payload, []byte("\n"))
-		// Only extract the requested stream type
-		if int(prefix[0]) == streamType {
+		// Discard payload if streamType doesn't match requested
+		if h.streamType == streamType {
+			lines := bytes.Split(payload, []byte("\n"))
 			for _, line := range lines {
 				if len(line) != 0 {
-					logging.Stderr("[filterdockerstream] read line from stream: ", string(line))
+					logging.Stderr("[FilterDockerstream] extracted line: %s", string(line))
 					result = append(result, string(line))
 				}
 			}
 		}
-		_, err = reader.Read(prefix)
-		if err != nil {
-			continue
+		h, err = readFrameHeader(reader)
+		if err == io.EOF {
+			break
 		}
 	}
-
 	return result, nil
 }
+
+// readFrameHeader returns a Docker stream frameHeader from the supplied Reader
+// See - https://docs.docker.com/engine/api/v1.29/#operation/ContainerAttach
+func readFrameHeader(reader io.Reader) (*frameHeader, error) {
+	logging.Stderr("[readFrameHeader] reading header")
+	prefix := make([]byte, stdWriterPrefixLen)
+
+	_, err := reader.Read(prefix)
+	if err != nil {
+		logging.Stderr("[readFrameHeader] error thrown reading frameHeader: %s", err)
+		return nil, err
+	}
+
+	header := frameHeader{
+		streamType:  int(prefix[0]),
+		payloadSize: int(binary.BigEndian.Uint32(prefix[stdWriterSizeIndex:])),
+	}
+	logging.Stderr("[readFrameHeader] extracted streamType: %d, payloadSize: %d", header.streamType, header.payloadSize)
+	return &header, nil
+}
+
+// readFrameHeader returns a payload byte array from the Reader whose length is
+// specified in the frameHeader
+func readFramePayload(r io.Reader, h *frameHeader) ([]byte, error) {
+	logging.Stderr("[readFramePayload] reading payloadSize: %d", h.payloadSize)
+	// payload := make([]byte, h.payloadSize)
+	lr := io.LimitReader(r, int64(h.payloadSize))
+	payload, err := ioutil.ReadAll(lr)
+	logging.Stderr("[readFramePayload] payload size: %d", len(payload))
+	if err != nil {
+		err = fmt.Errorf("[readFramePayload] Error reading payload: %s", err)
```

Comment:
`io.EOF` doesn't get returned by `ioutil.ReadAll` so we don't need to care about it in this case, any error returned is real, as with previous comment I'm adding more info to the comment in code